File: python/src/kinetic_mhd_pinn/module4_alfven_continuum_eigenmodes/solver.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from kinetic_mhd_pinn.common.scaffold import ScaffoldModule
from kinetic_mhd_pinn.common.state import CoupledState

from .benchmark import solve_benchmark
from .model import BenchmarkCase, BenchmarkResult, load_case


@dataclass
class AlfvenContinuumEigenmodesModule(ScaffoldModule):
    """Module 4 interface plus the first validated reduced benchmark.

    The shared-state ``run`` method preserves the architecture smoke test. The
    actual physics milestone is exposed through ``solve_case`` and implements a
    self-adjoint two-harmonic reduced shear-Alfven eigenvalue model.
    """

    module_id: str = "module_4"
    required_products: tuple[str, ...] = ()
    produced_marker: str | None = "module_4_contract_ready"

    def run(self, state: CoupledState) -> CoupledState:
        return super().run(state)

    @staticmethod
    def solve_case(case: BenchmarkCase | str | Path) -> BenchmarkResult:
        resolved = load_case(case) if isinstance(case, (str, Path)) else case
        return solve_benchmark(resolved)

File: matlab/src/+kinetic_mhd_pinn/+module4_alfven_continuum_eigenmodes/solver.m
function output = solver(caseInput)
%SOLVER Module 4 public entry point.
% With no input, returns the architecture contract for module_4. With a JSON
% path or decoded case struct, solves the reduced two-harmonic benchmark.
if nargin == 0
    output = kinetic_mhd_pinn.common.ScaffoldModule( ...
        'module_4', {}, 'module_4_contract_ready');
else
    output = kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes.solve_benchmark(caseInput);
end
end
